# Assignment 4: End-of-Game Film Analysis

**Description**: Break down the final moments of Knicks/Celtics playoff Game 1 from 2025. Recommend a better endgame strategy based on data.

1. How much was Jrue Holiday’s offensive rebound at 1:37 in the 4th quarter worth in Win Probability?
2. What would have been worth more for Boston’s win probability with 0:55 left in the 4th quarter - Jrue Holiday taking that layup or him passing to Al Horford for a corner 3?
3. Analyze New York’s clock management on the possession starting with 0:53 in the 4th quarter
4. How much win probability did New York gain with OG Anunoby drawing the And-1 foul with 3:25 remaining in Overtime?
5. When Boston gets the ball with 0:53 left in Overtime, is there any scenario they should go for a quick 2 instead of taking a 3? Why?
6. What other takeaways do you have from late game? Are there any other key inflection points where you would have recommended a different strategy?
-----

## Question 1
1. How much was Jrue Holiday’s offensive rebound at 1:37 in the 4th quarter worth in Win Probability?

Using the given NBA Win Probability Calculator [inpredictable](https://stats.inpredictable.com/nba/wpCalc.php), (Q4, 1:37, 1, Y) the **Win Probability** for the **Boston Celtics** was 70.8% after Jrue Holiday's offensive rebound with 1:37 remaining in the 4th quarter.

Had the New York Knicks grabbed the Defensive Rebound in this instance, the Boston Celtics' Win Probability would've dropped to 56.3% (Q4, 1:37, 1, N).

Therefore, Jrue Holiday's Offensive Rebound was worth 14.5% of Win Probability!

## Question 2
2. What would have been worth more for Boston’s win probability with 0:55 left in the 4th quarter - Jrue Holiday taking that layup or him passing to Al Horford for a corner 3?

In [2]:
# Define Functions
def calculate_probability_of_made_shot(offensive_player_name, defensive_player_name,
                                       offensive_player, defensive_player, 
                                       offensive_quality = 0.6, defensive_quality = 0.4,
                                       zone = None
                                      ):
    """
    Parameters:
        offensive_player_name = The Name of the Offensive Player
        offensive_player = The FG% of Making the Shot in a Specific Zone
        offensive_skill = Overall Weight of the Quality of the Shot, Defaults to 60%

        defensive_player_name = The Name of the Defensive Player
        defensive_player = The FG% of Opponents Making a Shot in a Specific Zone
        defensive_skill = Overall Weight of the Quality of the Shot, Defaults to 40%

        zone = The Area of the Court Where the Analysis is Conducted
    
    Returns:
        The Probability that the Offensive Player Makes the Shot Against the Defensive Player
    """
    
    probability = offensive_player*offensive_quality + defensive_player*defensive_quality

    if zone is not None:
        print(offensive_player_name + " has a " + str(round(probability * 100, 2)) + "% chance to make the shot in the " + zone + " against " + defensive_player_name)
        return probability

    else:
        print(offensive_player_name + " has a " + str(round(probability * 100, 2)) + "% chance to make the shot against " + defensive_player_name)

def expected_points(probability, potential_points):
    """
    Parameters:
        probability = 
        potential_points = 
    Returns:
    
    """
    return probability*potential_points

In [3]:
# According to NBA.com, Jrue Holiday made 93-out-of-132 FGAs in the Restricted Area
# https://www.nba.com/stats/players/shooting?Season=2024-25&TeamID=1610612738&PerMode=Totals&DistanceRange=By+Zone
holiday_layup = 93/132 

# According to NBA.com, players made 293-out-of-446 FGAs less than 6FT away from the basket while Karl-Anthony Towns was defending
# https://www.nba.com/stats/players/defense-dash-lt6?PerMode=Totals&Season=2024-25
towns_layup_defense = 293 / 446

# Reviewing the footage, Jrue Holiday was slightly covered at the time of release. He had a great angle and a great shot-view.
# We'll give Jrue Holiday an 80% Shot Quality, and we'll give Karl-Anthony Towns a 20% Shot Defense

question_2_holiday = calculate_probability_of_made_shot("Jrue Holiday", "Karl-Anthony Towns",
                                                        holiday_layup, towns_layup_defense,
                                                        offensive_quality = 0.8, defensive_quality = 0.2,
                                                        zone = "Restricted Area"
                                                       )


# According to NBA.com, Al Horford made 22-out-of-62 Right Corner 3s
# https://www.nba.com/stats/players/shooting?Season=2024-25&TeamID=1610612738&PerMode=Totals&DistanceRange=By+Zone
horford_three = 22/62

# According to NBA.com, players made 115-out-of-320 3FGAs while Jalen Brunson was defending
# https://www.nba.com/stats/players/defense-dash-3pt
brunson_three_defense = 115/320

# Reviewing the footage, the Defensive Rotation would've forced Jalen Brunson to cover the distance from the Paint-to-the-Right-Corner
# We'll give Al Horford a 95% Shot Quality, and we'll give Jalen Brunson a 5% Shot Defense
question_2_horford = calculate_probability_of_made_shot("Al Horford", "Jalen Brunson",
                                                        horford_three, brunson_three_defense,
                                                        offensive_quality = 0.95, defensive_quality = 0.05,
                                                        zone = "Right Corner 3"
                                                       )

Jrue Holiday has a 69.5% chance to make the shot in the Restricted Area against Karl-Anthony Towns
Al Horford has a 35.51% chance to make the shot in the Right Corner 3 against Jalen Brunson


Using the values we calculated on the probability of each shot being made, we can input these values into [inpredictable](https://stats.inpredictable.com/nba/wpCalc.php). 

In the actual game, there were 53.4 seconds remaining at the end of Jrue Holiday's made layup. We will assume that there will be 53 seconds remaining in all outcomes (made basket, missed bakset).

In addition, in the event of both players missing, there are 5 NYK players within the 3PT line. Thus, the likelihood of an Offensive Rebound for the Boston Celtics is extremely low. We will ignore this case for this question. Therefore, in the event of a missed shot, New York will grab the Defensive Rebound

In [5]:
# Win Probability
def win_probability(make, win_make,
                    miss, win_miss
                   ):
    """
    Parameters:

    Returns:
    
    """
    return make*win_make + miss*win_miss
###########

# Jrue Holiday Win Probability
holiday_make_event = question_2_holiday
win_prob_holiday_make = 0.408 # Derived from Inpredictable, Q4, 0:53, 0, N

holiday_miss_event = 1 - holiday_make_event
win_prob_holiday_miss = 0.137 # Derived from Inpredictable, Q4, 0:53, -2, N

holiday_wp = round(100*win_probability(holiday_make_event, win_prob_holiday_make, holiday_miss_event, win_prob_holiday_miss), 2)


# Al Horford Win Probability
horford_make_event = question_2_horford
win_prob_horford_make = 0.587 # Derived from Inpredictable, Q4, 0:53, 1, N

horford_miss_event = 1 - horford_make_event
win_prob_horford_miss = 0.137 # Derived from Inpredictable, Q4, 0:53, -2, N

horford_wp = win_probability(horford_make_event, win_prob_horford_make,
                             horford_miss_event, win_prob_horford_miss
                            )
horford_wp = round(100*horford_wp, 2)

print("Jrue Holiday's Win Probability: " + str(holiday_wp) + "%")
print("Al Horford's Win Probability: " + str(horford_wp) + "%")

Jrue Holiday's Win Probability: 32.54%
Al Horford's Win Probability: 29.68%


From our calculation above, we found that Jrue Holiday's layup attempt going for two points had a **win probability** of 32.54%. We also found that if Jrue Holiday passed to Al Horford in the right corner, Horford's three-point attempt would have had a **win probability** of 29.68%. 

Therefore, it was more advantageous for the Boston Celtics to take the layup as opposed to the three.

## Question 3

3. Analyze New York’s clock management on the possession starting with 0:53 in the 4th quarter

New York's clock management was phenomenal with 53 seconds left in regulation. Jalen Brunson took his time getting past half court, and he didn't start the action until there were 12 seconds remaining in the Shot Clock and 42 seconds remaining in the Game Clock. Brunson and his team were well aware of the time in the game, and they paced themselves to go for a two-for-one opportunity to give themselves a shot at potentially the last shot of the game. 

However, their shot selection at the end of possession was poorly made. An Anunoby contested midrange over Jayson Tatum is not the type of shot we should be aiming for in these type of situations. His open three would've been a better shot at 34 seconds in regulation, as I will detail below. 

For simplicity, we'll assume the Boston Celtics grab the Defensive Rebound in the case of both misses as at the time of both shots, the majority of players inside the 3PT-Arch were Celtics players.

In terms of the **Time Remaining**, after the hypothetical 3PT Shot, we'll assume there was 32 seconds remaining in the game clock at the time of the make or rebound after the miss. After the midrange shot, we see that there was 29 seconds remaining in the game clock.

In [7]:
# According to NBA.com, OG Anunoby made 89-out-of-243 Above the Break 3s
# https://www.nba.com/stats/players/shooting?Season=2024-25&TeamID=1610612752&PerMode=Totals&DistanceRange=By+Zone
anunoby_three = 89 / 243

# According to NBA.com, players made 108-out-of-331 3PA while Jayson Tatum was defending
# https://www.nba.com/stats/players/defense-dash-3pt
tatum_three_defense = 108 / 331

# Reviewing the footage, had OG Anunoby attempted the Above the Break 3, he was slighltly off-balance, and Jayson Tatum was slightly late in his Defensive Rotation.
# We'll give OG Anunoby a 70% Shot Quality, and we'll give Jayson Tatum a 30% Shot Defense

anunoby_three_prob = calculate_probability_of_made_shot("OG Anunoby", "Jayson Tatum",
                                                        anunoby_three, tatum_three_defense,
                                                        offensive_quality = 0.7, defensive_quality = 0.3,
                                                        zone = "Above the Break 3"
                                                       )


# According to NBA.com, OG Anunoby made 25-out-of-78 shots from Midrange
# https://www.nba.com/stats/players/shooting?Season=2024-25&TeamID=1610612752&PerMode=Totals&DistanceRange=By+Zone
anunoby_mid = 25 / 78

# According to NBA.com, players made 113-out-of-323 FGAs from >15 from-the-basket while Jayson Tatum was defending
# https://www.nba.com/stats/players/defense-dash-gt15
tatum_mid_defense = 113 / 323

# Reviewing the footage, OG Anunoby's Midrange Shot was Off-Balance and Smothered by Jayson Tatum's Amazing Shot Contest
# We'll give OG Anunoby a 10% Shot Quality, and we'll give Jayson Tatum a 90% Shot Defense

anunoby_mid_prob = calculate_probability_of_made_shot("OG Anunoby", "Jayson Tatum",
                                                        anunoby_mid, tatum_mid_defense,
                                                        offensive_quality = 0.1, defensive_quality = 0.9,
                                                        zone = "Above the Break 3"
                                                       )

OG Anunoby has a 35.43% chance to make the shot in the Above the Break 3 against Jayson Tatum
OG Anunoby has a 34.69% chance to make the shot in the Above the Break 3 against Jayson Tatum


In [8]:
# OG Anunoby - 3PT Win Probability
anunoby_make_3_event = anunoby_three_prob
win_prob_anunoby_make_3 = 0.897 # Derived from Inpredictable

anunoby_miss_3_event = 1 - anunoby_make_3_event
win_prob_anunoby_miss_3 = 0.393 # Derived from Inpredictable

anunoby_3_wp = round(100*win_probability(anunoby_make_3_event, win_prob_anunoby_make_3, 
                                       anunoby_miss_3_event, win_prob_anunoby_miss_3), 2)


# OG Anunoby - 2PT Win Probability
anunoby_make_2_event = anunoby_mid_prob
win_prob_anunoby_make_2 = 0.791 # Derived from Inpredictable: Q4, 0:29, 2, N

anunoby_miss_2_event = 1 - anunoby_make_2_event
win_prob_anunoby_miss_2 = 0.371 # Derived from Inpredictable: Q4, 0:29, 0, N

anunoby_2_wp = round(100*win_probability(anunoby_make_2_event, win_prob_anunoby_make_2, 
                                       anunoby_miss_2_event, win_prob_anunoby_miss_2), 2)


print("OG Anunoby's 3PT Win Probability: " + str(anunoby_3_wp) + "%")
print("OG Anunoby's 2PT Win Probability: " + str(anunoby_2_wp) + "%")

OG Anunoby's 3PT Win Probability: 57.15%
OG Anunoby's 2PT Win Probability: 51.67%


As shown above, OG Anunoby's hypothetical 3PT Shot with 34 seconds remaining in the game would've increased the New York Knicks' probability of winning by 5.48%!

## Question 4

4. How much win probability did New York gain with OG Anunoby drawing the And-1 foul with 3:25 remaining in Overtime?

At the start of the possession, after the Boston Celtics' turnover at 3:31, the New York Knicks' **Win Probability** was at 55.0% (OT, 3:31, 0, Y). Below we will calculate the New York Knicks' Win Probability before OG Anunoby's Free Throw Attempt on the And-1 Foul, and we will calculate the New York Knicks' Win Probability after the made And-1 Foul Shot.

In [10]:
# According to basketball-reference.com, OG Anunoby is a Career 76.1% from the Free Throw Line
# https://www.basketball-reference.com/players/a/anunoog01.html
anunoby_free_throw = 0.761

# Rebound Ratings of the Knicks and Celtics
knicks_orb = 0.26
knicks_opp_drb = 1 - 0.26
celtics_drb = 0.76
celtics_opp_orb = 1 - 0.76

# Calculate the Probability of Knicks Offensive Rebound
prob_knicks_oreb = knicks_orb / (knicks_orb + celtics_drb)
prob_celtics_dreb = celtics_drb / (knicks_orb + celtics_drb)

print("Probability of Knicks OREB: " + str(round(100*prob_knicks_oreb, 2)) + "%")
print("Probability of Celtics DREB: " + str(round(100*prob_celtics_dreb, 2)) + "%")

############## Three Events ##############

# First Event: OG Anunoby Makes the And-1 Foul Shot
anunoby_free_throw = 0.761 # From Basketball Reference
made_free_throw_wp = 0.705 # OT, 3:25, 3, N

# Second Event: OG Anunoby Misses the And-1 Foul Shot, Knicks Grab OREB
anunoby_miss_free_throw = 1 - anunoby_free_throw
prob_knicks_oreb = knicks_orb / (knicks_orb + celtics_drb)
miss_free_throw_knicks_reb = 0.707 # OT, 3:25, 2, Y

# Third Event: OG Anunoby Misses the And-1 Foul Shot, Celtics Grab DREB
anunoby_miss_free_throw = 1 - anunoby_free_throw
prob_celtics_dreb = celtics_drb / (knicks_orb + celtics_drb)
miss_free_throw_celtics_reb = 0.625 # OT, 3:25, 2, N


###### Calculate the Win Probability After the Foul and Before the Free Throw Shot was Taken
nyk_wp_q4 = anunoby_free_throw*made_free_throw_wp + anunoby_miss_free_throw*prob_knicks_oreb*miss_free_throw_knicks_reb + anunoby_miss_free_throw*prob_celtics_dreb*miss_free_throw_celtics_reb
nyk_wp_q4 = round(nyk_wp_q4,3)
print("")
print("New York Knicks' Win Probability at the Start of the Possession: 55.0%") # OT, 3:31, 0, Y
print("New York Knicks' Win Probability Before the Free Throw Attempt: " + str(100*nyk_wp_q4) + "%")
print("New York Knicks' Win Probability After the Made Free Throw Attempt: 70.5%") # OT, 3:25, 3, N

Probability of Knicks OREB: 25.49%
Probability of Celtics DREB: 74.51%

New York Knicks' Win Probability at the Start of the Possession: 55.0%
New York Knicks' Win Probability Before the Free Throw Attempt: 69.1%
New York Knicks' Win Probability After the Made Free Throw Attempt: 70.5%


Our calculation above showed that that OG Anunoby's And-1 play increased the New York Knicks' Win Probability by 14.1%. In addition, it increased even further by 1.4% after the made foul shot!

## Question 5
5. When Boston gets the ball with 0:53 left in Overtime, is there any scenario they should go for a quick 2 instead of taking a 3? Why?

When Jaylen Brown goes for a layup with 43 seconds left in Overtime, he attempts it in the restricted area against Karl-Anthony Towns. Unlike in Question 2, where Towns was a little late on the rotation, he is right there in a perfect, vertical contest.


In [12]:
# According to NBA.com, Jaylen Brown made 206-out-of-305 FGAs in the Restricted Area
# https://www.nba.com/stats/players/shooting?Season=2024-25&TeamID=1610612738&PerMode=Totals&DistanceRange=By+Zone
brown_layup = 206 / 305

# According to NBA.com, players made 293-out-of-446 FGAs less than 6FT away from the basket while Karl-Anthony Towns was defending
# https://www.nba.com/stats/players/defense-dash-lt6?PerMode=Totals&Season=2024-25
towns_layup_defense = 293 / 446

# Reviewing the footage, Jaylen Brown was heavily contested by Karl-Anthony Towns as Towns made a perfect rotation to cover.
# We'll give Jaylen Brown a 10% Shot Quality, and we'll give Karl-Anthony Towns a 90% Shot Defense
question_5_brown = calculate_probability_of_made_shot("Jaylen Brown", "Karl-Anthony Towns",
                                                      brown_layup, towns_layup_defense,
                                                      offensive_quality = 0.1, defensive_quality = 0.9,
                                                      zone = "Restricted Area"
                                                     )

Jaylen Brown has a 65.88% chance to make the shot in the Restricted Area against Karl-Anthony Towns


However, spotting up in the corner is the same player we analyzed in Question 2 - Al Horford. Since Towns rotated off of Al Horford, Horford is the logical player to take a 3PA in this possession. Below, we will calculate had Jaylen Brown made a *drive-and-kick* play.

Reviewing the game, Josh Hart would be responsible for rotating, as he lost his man on the drive, but he was not in an ideal position to rotate properly. Nevertheless, the passing angle from Jaylen Brown wasn't great either, so we'll offset both of these scenarios. See the comments in the calculation. 

In the event of Al Horford's hypothetical 3PA: 
- If Horford makes the 3PA, then we will assume 40 seconds remain in Overtime
- If Horford misses the 3PA, then we will assume New York grabs the Defensive Rebound because there are 3 Knicks players in great position to rebound the ball. 

In [14]:
# According to NBA.com, Al Horford made 22-out-of-62 Right Corner 3s
# https://www.nba.com/stats/players/shooting?Season=2024-25&TeamID=1610612738&PerMode=Totals&DistanceRange=By+Zone
horford_three = 22/62

# According to NBA.com, players made 168-out-of-458 3FGAs while Josh Hart was defending
# https://www.nba.com/stats/players/defense-dash-3pt
hart_three_defense = 168 / 458

# Reviewing the footage, the Defensive Rotation would've forced Josh Hart to cover the distance from the Paint-to-the-Right-Corner.
# However, the passing lane and angle might've been tight, so that might've given a good amount of time for Hart to recover.
# We'll give Al Horford a 90% Shot Quality, and we'll give Josh Hart a 10% Shot Defense
question_5_horford = calculate_probability_of_made_shot("Al Horford", "Josh Hart",
                                                        horford_three, hart_three_defense,
                                                        offensive_quality = 0.90, defensive_quality = 0.10,
                                                        zone = "Right Corner 3"
                                                       )

# Jaylen Brown Win Probability
brown_make_event = question_5_brown
win_prob_brown_make = 0.231 # OT, 0:42, -1, N

brown_miss_event = 1 - question_5_brown
win_prob_brown_miss = 0.054 # OT, 0:42, -3, N

brown_wp = round(100*win_probability(brown_make_event, win_prob_brown_make,
                                     brown_miss_event, win_prob_brown_miss), 2)
                                


# Al Horford Win Probability
horford_make_event = question_5_horford
win_prob_horford_make = 0.397 # Derived from Inpredictable, OT, 0:40, 0, N

horford_miss_event = 1 - horford_make_event
win_prob_horford_miss = 0.052 # Derived from Inpredictable, OT, 0:40, -3, N

horford_wp = win_probability(horford_make_event, win_prob_horford_make,
                             horford_miss_event, win_prob_horford_miss
                            )
horford_wp = round(100*horford_wp, 2)
print("")
print("Jaylen Brown's Win Probability: " + str(holiday_wp) + "%")
print("Al Horford's Win Probability: " + str(horford_wp) + "%")

Al Horford has a 35.6% chance to make the shot in the Right Corner 3 against Josh Hart

Jaylen Brown's Win Probability: 32.54%
Al Horford's Win Probability: 17.48%


In this possession, Jaylen Brown's layup attempt against Karl-Anthony Towns was the best shot. However, it was early in the shot clock. Getting ahead to **Question 6**, when Jaylen Brown drove the ball on Josh Hart, he had beaten him on the killer crossover. Brown had ample time to push Hart *in jail*, and attempt to do a floater before Towns had time to rotate.

Overall, it was still a good shot. Brown took it to the hole, and he drew an unlucky bounce off the rim. 

## Question 6

6. What other takeaways do you have from late game? Are there any other key inflection points where you would have recommended a different strategy?

The end of regulation execution from the New York Knicks was phenomenal. With 5 seconds remaining, Tom Thibbedau drew up an amazing backdoor cut play where Brunson's floater was just out-of-reach of the out-stretched arms of Horford. 

Even if we conservatively say that Al Horford was playing smothering defense and give him a Defensive Shot Quality of 90%, although, I think it's more on the lines of 50%, Jalen Brunson had a 60.14% chance to make the shot to win the game. 

In [16]:
# Calculate Probability of Making the Game Winner in Q4
brunson_layup = 145 / 226
horford_layup_defense = 154 / 258

brunson_game_winner = calculate_probability_of_made_shot("Jalen Brunson", "Al Horford",
                                                         brunson_layup, horford_layup_defense,
                                                         offensive_quality = 0.1, defensive_quality = 0.9,
                                                         zone = "Restricted Area"
                                                        )


Jalen Brunson has a 60.14% chance to make the shot in the Restricted Area against Al Horford


On the Boston Celtics' side, I think they could've done a better job in the last possession execution. Only one screen was set, and there was no movement to get open. It allowed New York to strip the ball and prevent even a Shot Attempt. In addition, I'm not sure what bringing the ball down low in that situation would bring when Brown caught the ball. It exposed the ball too much to a long and defensive-beast in Mikal Bridges. A cross-court pass was too dangerous in that situation. I would've liked to see a curl screen toward the ball instead of flaring out. Even at the end of regulation, Boston tried a post-up, fadeaway, contested three that went nowhere. An alleyoop layup might've had a better chance than that play call. In this situations, I'm reminded of Paul Westphal's end-of-game play call to win against Portland where Oliver Miller inbounded against the backboard and Charles Barkley grabbed the ball from the floor for quick layup to win the game. 

Overall, New York played amazing defense to close out Game 1 in dramatic fashion. I believe Boston's Possession with 4:26 remaining in Overtime sums up New York's playstyle throughout the end of the game: suffocating and unrelating defense. They denied Boston of any momentum, and they positioned themselves for an 8-2 run in which they did not look back.